libs/datamodel/core/src/transform/ast_to_dml/db/attributes.rs

```diff
@@ -542,13 +657,108 @@ fn model_index<'ast>(
+
+    let (name, db_name) = if ctx.db.preview_features.contains(NamedConstraints) {
+        let generated_name = ConstraintNames::index_name(
+            model_db_name,
+            &field_db_names,
+            IndexType::Normal,
+            ctx.db.active_connector(),
+        );
+
+        let db_name = match args.optional_arg("map").map(|name| name.as_str()) {
+            Some(Ok("")) => error_on_empty_string_cow(args, ctx, "map"),
+            Some(Ok(name)) => Some(name.into()),
+            Some(Err(err)) => push_error_cow(ctx, err),
+            None => None,
+        };
+
+        validate_db_name(&ast_model, args, ctx, &db_name, "@@index");
+
+        // this is for compatibility reasons
+        let db_name = match (name, db_name) {
+            (Some(_), Some(map)) => Some(map), //todo error here
+            //backwards compatibility, accept name arg on normal indexes and use it as map arg
+            (Some(name), None) => Some(name.into()),
+            (None, Some(map)) => Some(map),
+            (None, None) => Some(generated_name.into()),
+        };
+
+        (None, db_name)
+    } else {
+        //old default name logic moved from sql schema calculator
+
+        let separator = if ctx.is_sql_server() { "_" } else { "." };
+        let generated_name = format!(
+            "{table}{separator}{fields}_index",
+            table = &model_db_name,
+            separator = separator,
+            fields = field_db_names.join("_")
+        );
+
+        (None, name.map(|f| f.into()).or_else(|| Some(generated_name.into())))
+    };
+
+    index_data.name = name;
+    index_data.db_name = db_name;
 
     data.indexes.push((args.attribute(), index_data));
 }
 
+fn get_field_db_names<'ast>(model_id: ModelId, fields: &[FieldId], ctx: &mut Context<'ast>) -> Vec<&'ast str> {
+    fields
+        .iter()
+        .map(|&field_id| {
+            ctx.db
+                .get_field_database_name(model_id, field_id)
+                .unwrap_or(&ctx.db.ast[model_id][field_id].name.name)
+        })
+        .collect()
+}
+
+fn get_name_argument<'ast>(args: &mut Arguments<'ast>, ctx: &mut Context<'ast>) -> Option<&'ast str> {
+    match args.optional_arg("name").map(|name| name.as_str()) {
+        Some(Ok("")) => error_on_empty_string(args, ctx, "name"),
+        Some(Ok(name)) => Some(name),
+        Some(Err(err)) => push_error(ctx, err),
+        None => None,
+    }
+}
+
+fn get_map_argument<'ast>(
+    args: &mut Arguments<'ast>,
+    ctx: &mut Context<'ast>,
+    generated_name: String,
+) -> Option<Cow<'ast, str>> {
+    match args.optional_arg("map").map(|name| name.as_str()) {
+        Some(Ok("")) => error_on_empty_string_cow(args, ctx, "map"),
+        Some(Ok(name)) => Some(name.into()),
+        Some(Err(err)) => push_error_cow(ctx, err),
+        None => Some(generated_name.into()),
+    }
+}
+
+fn push_error<'ast>(ctx: &mut Context<'ast>, err: DatamodelError) -> Option<&'ast str> {
+    ctx.push_error(err);
+    None
+}
```

Comment:
I don't really understand this function and the one below. It looks like they always return `None`?

Reply:
Ah I get it now, it's so the type of the error branches matches up with the type of the main branch. I think the `generated_name` in `get_map_argument` feels a bit out of place. We could either `get_map_argument(args, ctx).or_else(|| generate_default_name(...))` for example. But it's a nitpick, feel free to ignore if it delays too much.

Reply:
But this would always insert a generated name instead of only in cases where there was no error. I'm not yet sure which behaviour would be better.